sdk/identity/azure-identity/src/managed_identity_credential.cpp

```diff
@@ -49,16 +51,38 @@ ManagedIdentityCredential::ManagedIdentityCredential(
     Azure::Core::Credentials::TokenCredentialOptions const& options)
     : TokenCredential("ManagedIdentityCredential")
 {
-  m_managedIdentitySource = CreateManagedIdentitySource(GetCredentialName(), clientId, {}, options);
+  m_managedIdentitySource
+      = CreateManagedIdentitySource(GetCredentialName(), clientId, {}, {}, options);
 }
 
 ManagedIdentityCredential::ManagedIdentityCredential(
-    ResourceIdentifier const& resourceId,
-    Azure::Core::Credentials::TokenCredentialOptions const& options)
+    ManagedIdentityCredentialOptions const& options)
     : TokenCredential("ManagedIdentityCredential")
 {
-  m_managedIdentitySource
-      = CreateManagedIdentitySource(GetCredentialName(), {}, resourceId.ToString(), options);
+  int numOptionsSet = 0;
+  if (!options.ClientId.empty())
+  {
+    numOptionsSet++;
+  }
+  if (!options.ObjectId.empty())
+  {
+    numOptionsSet++;
+  }
+  if (!options.ResourceId.ToString().empty())
+  {
+    numOptionsSet++;
+  }
+  if (numOptionsSet > 1)
+  {
+    throw std::invalid_argument("Only one of ClientId, ObjectId, or ResourceId can be set in "
+                                "ManagedIdentityCredentialOptions.");
+  }
```

Comment:
Needing this validation check is the downside of this design choice.

The user could accidentally provide more than one id in the options, so we'd want to validate that.

The existing design in https://github.com/Azure/azure-sdk-for-cpp/pull/5910 doesn't have that downside.